File: ios/Sources/WattRouter/Inference.swift
// Inference.swift — asking a model, and reading its answer as it is produced.
//
// History
//   2026-08-07  A. Sigdel  Created.
//
// Contents
//   InferenceError     Why an attempt failed, and whether another model is worth trying.
//   Inference          Ask one model; receive the answer in chunks.
//   ScriptedInference  A fixed answer, delivered in chunks. For previews and tests.
//
// Streaming rather than a single return, and that decision is the whole shape of
// this file. Both implementations that will exist produce a token at a time — the
// upstream API over the network, a local model in this process — and a
// `-> String` signature would force each of them to buffer. It would also make
// the latency split unmeasurable: prompt-eval and generation are distinguishable
// only by when the first chunk arrives. Retrofitting streaming means changing
// every caller, so it goes in before there is one.
//
// The other half is the error type. A chain exists so a failed attempt can be
// retried against the next model, and the walk that does that lives above this
// seam rather than inside it. So the distinction `upstream.rs:146` makes in a
// match arm has to be a value here: a server error is worth another model, a
// client error is not, "since the next model would reject the same body
// identically".

import Foundation

/// Why an attempt failed.
///
/// The cases exist to be told apart by a caller walking a chain, which is why
/// this is not one case with a string in it.
public enum InferenceError: Error, Equatable, Sendable {
    /// The model could not be reached, or failed on its own account: a dropped
    /// connection, a timeout, a 5xx. Another model may answer the same request.
    case unavailable(model: String, detail: String)

    /// The request was rejected on its merits — a 4xx. Every model behind the
    /// same API would reject it identically, so a chain stops here rather than
    /// spending its remaining attempts proving that.
    case rejected(model: String, status: Int, detail: String)

    /// Every model in the chain failed. Produced by the walk over a chain, never
    /// by a single attempt; an [`Inference`] must not return it.
    case exhausted(tried: Int, last: String)

    /// Whether the next model in a chain is worth trying.
    ///
    /// The chain walk reads this instead of matching, so the rule is written once
    /// here and a new case cannot silently default to retrying.
    public var isWorthAnotherModel: Bool {
        switch self {
        case .unavailable: true
        case .rejected, .exhausted: false
        }
    }
}

/// One model, asked one question.
///
/// Deliberately below the chain: an implementation speaks to a single model and
/// reports how it failed, and choosing what to try next is somebody else's job.
public protocol Inference: Sendable {
    /// Ask `model` to continue `conversation`.
    ///
    /// - Parameters:
    ///   - conversation: everything said so far, oldest first.
    ///   - model: the model to ask, named as the provider names it. A [`Step`]
    ///     carries one, along with where it runs.
    ///   - maxTokens: a cap on the reply, or `nil` for the provider's default.
    /// - Returns: the answer in the order it is produced. Chunks are fragments of
    ///   text and carry no token boundary — a caller that needs the whole answer
    ///   concatenates them.
    ///
    /// # Rely
    /// A failure before the first chunk is the caller's to retry. A failure after
    /// one has been yielded is not: text has already reached the caller, and no
    /// second model can un-deliver it. Conformances must therefore not yield a
    /// chunk they are not committed to — in particular, nothing may be emitted
    /// before the response status is known.
    ///
    /// # Atomic
    /// Safe to call concurrently. Each call owns its own stream; ending one, by
    /// cancellation or by leaving the loop early, must not disturb another.
    func complete(_ conversation: Conversation, model: String, maxTokens: Int?)
        -> AsyncThrowingStream<String, any Error>
}

/// A fixed answer, delivered in chunks, optionally ending in a failure.
///
/// In the library rather than the test target on purpose: a SwiftUI preview needs
/// something to render, and so does the app before it can reach a model at all.
public struct ScriptedInference: Inference {
    /// Yielded in order, then `failure` if there is one.
    public let chunks: [String]
    /// Waited before each chunk, so a caller can be seen to receive them
    /// separately rather than all at once.
    public let perChunk: Duration
    /// Ends the stream after the last chunk, or `nil` to end it normally.
    public let failure: InferenceError?

    /// Script the chunks exactly.
    ///
    /// Empty `chunks` with a `failure` is a model that fails outright — the case
    /// a chain walk is allowed to retry.
    public init(chunks: [String], perChunk: Duration = .zero, failure: InferenceError? = nil) {
        self.chunks = chunks
        self.perChunk = perChunk
        self.failure = failure
    }

    /// Deliver `answer` a word at a time, as a model would.
    ///
    /// The spaces are kept on the chunks rather than dropped, so concatenating
    /// what arrives reproduces `answer` and a caller can join without knowing it
    /// was ever split.
    public init(answer: String, perChunk: Duration = .zero) {
        let words = answer.split(separator: " ", omittingEmptySubsequences: false)
        self.init(
            chunks: words.enumerated().map { index, word in
                index == words.count - 1 ? String(word) : "\(word) "
            },
            perChunk: perChunk)
    }

    public func complete(_ conversation: Conversation, model: String, maxTokens: Int?)
        -> AsyncThrowingStream<String, any Error>
    {
        AsyncThrowingStream { continuation in
            let task = Task {
                do {
                    for chunk in chunks {
                        // Zero is the common case — a test that only wants the
                        // text — and sleeping for it would still suspend.
                        if perChunk > .zero { try await Task.sleep(for: perChunk) }
                        continuation.yield(chunk)
                    }
                    continuation.finish(throwing: failure)
                } catch {
                    // Cancellation arrives here, from the sleep.
                    continuation.finish(throwing: error)
                }
            }
            // A caller that leaves its loop early ends the stream; without this
            // the task would run to the end of the script regardless, which for a
            // real implementation is a request nobody is waiting for.
            continuation.onTermination = { _ in task.cancel() }
        }
    }
}

File: android/core/src/main/kotlin/com/getlora/wattrouter/Inference.kt
// Inference.kt: asking a model, and reading its answer as it is produced.
//
// History
//   2026-08-08  A. Sigdel  Created.
//
// Contents
//   InferenceError     Why an attempt failed, and whether another model helps.
//   StreamEvent        What a model produces, in the order it produces it.
//   Inference          Ask one model; receive the answer in pieces.
//   ScriptedInference  A fixed answer in pieces. For tests and previews.
//
// Streaming rather than a single return, and that decides the shape of the
// file. A turn that shows nothing for twenty seconds is a turn people kill, and
// a `suspend fun ask(): String` would force every implementation to buffer.
// Retrofitting it means changing every caller, so it goes in before there is
// one.
//
// The other half is the error type. A chain exists so a failed attempt can be
// retried against the next model, and the walk doing that sits above this seam
// rather than inside it, so the distinction upstream.rs makes in a match arm
// has to be a value here. A server error is worth another model; a client error
// is not, because the next model would reject the same body identically.
//
// A Flow rather than a channel or a callback. It is cold, so nothing is
// requested until something collects; it cancels with its collector, which is
// what makes interrupting a turn a matter of cancelling a coroutine; and a
// failure arrives as an exception at the collection point rather than as a
// second parameter every caller has to remember to read.

package com.getlora.wattrouter

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow

/**
 * Why an attempt failed.
 *
 * The cases exist to be told apart by a caller walking a chain, which is why
 * this is not one class with a string in it.
 */
sealed class InferenceError(message: String) : Exception(message) {

    /**
     * The model could not be reached, or failed on its own account: a dropped
     * connection, a timeout, a 5xx. Another model may answer the same request.
     */
    class Unavailable(val model: String, val detail: String) :
        InferenceError("$model is unavailable: $detail")

    /**
     * The request was rejected on its merits, a 4xx. Every model behind the
     * same API would reject it identically, so a chain stops here rather than
     * spending its remaining attempts proving that.
     */
    class Rejected(val model: String, val status: Int, val detail: String) :
        InferenceError("$model rejected the request ($status): $detail")

    /**
     * Every model in the chain failed.
     *
     * Produced by the walk over a chain, never by a single attempt; an
     * [Inference] must not throw it.
     */
    class Exhausted(val tried: Int, val last: String) :
        InferenceError("all $tried models failed; the last said: $last")

    /**
     * Whether the next model in a chain is worth trying.
     *
     * The walk reads this rather than matching, so the rule is written once and
     * a case added later cannot silently default to retrying.
     */
    val isWorthAnotherModel: Boolean
        get() = this is Unavailable
}

/**
 * What a model produces, in the order it produces it.
 *
 * Two cases, and it will grow. There is deliberately no `Finished(reason)`: a
 * finish reason describes the stream that just ended, and every caller here
 * either has the pieces already or is about to be told by the stream closing.
 */
sealed interface StreamEvent {
    /** A fragment of the answer. Not a token, not a line: whatever arrived. */
    data class Text(val text: String) : StreamEvent

    /** A tool the model wants run, assembled from however many fragments. */
    data class Call(val call: ToolCall) : StreamEvent
}

/**
 * One model, asked one question.
 *
 * Not a `fun interface`: `maxTokens` has a default, and a functional interface
 * may not give its single method one. The default is worth more than the SAM
 * conversion, because every call site but one leaves it out.
 */
interface Inference {
    /**
     * Ask [model] to continue [conversation].
     *
     * The flow is cold: nothing is sent until it is collected, and cancelling
     * the collector cancels the request. Failure arrives as an [InferenceError]
     * thrown at the collection point.
     *
     * # Rely
     * The caller decides retries. This asks one model once, and must not throw
     * [InferenceError.Exhausted]; that belongs to whatever walks a chain.
     *
     * Nothing may be emitted before the response status is known. A caller that
     * has seen one event treats the attempt as delivered and will not retry it
     * against another model, so an event emitted early forecloses the retry
     * that would have worked.
     *
     * @param tools the provider's `tools` array, already JSON, or null for
     *   none. A parameter rather than something the conversation carries: #319
     *   records what happens when sending it is left to a later change.
     */
    fun complete(
        conversation: Conversation,
        model: String,
        tools: String? = null,
        maxTokens: Int? = null,
    ): Flow<StreamEvent>
}

/**
 * A fixed answer, delivered in pieces.
 *
 * In the library rather than the test source set so that a Compose preview and
 * a test share one, and so a screen can be looked at without a network.
 */
class ScriptedInference(
    private val events: List<StreamEvent>,
    private val failWith: InferenceError? = null,
) : Inference {

    /** Every model this was asked for, in order. What a chain walk is judged by. */
    val asked: List<String> get() = _asked
    private val _asked = mutableListOf<String>()

    constructor(says: String) : this(listOf(StreamEvent.Text(says)))

    override fun complete(
        conversation: Conversation,
        model: String,
        tools: String?,
        maxTokens: Int?,
    ): Flow<StreamEvent> = flow {
        // Recorded inside the flow, not outside it, because the flow is cold:
        // recording at construction would count a model nobody collected.
        _asked += model
        failWith?.let { throw it }
        events.forEach { emit(it) }
    }
}
